### PR TITLE
Validations require input via kwargs

### PR DESCRIPTION
**Pull request by gcheadle-vmware**
Contributes to #642 

## Feature Work

- [x] (@vmunishwar) allow assertion functions to be intuitively combined in boolean expressions — assertion functions return `True`, not `None`
- [x] (@pivotaljohn) support concise use of assertion functions in custom rules (e.g. `("custom def message", assert.min(1))` instead of `("custom def message", assert.min(1).check)`) — allow not just Callables, but also objects that have an `check` attribute that is a Callable.
- [x] implement `assert.not_null(v)` short-hand.
- [x] (@vmunishwar) store `min_len` and `max_len` as `starlark.Int`s (details: https://github.com/vmware-tanzu/carvel-ytt/pull/677#discussion_r910447266)
  - [x] remove comment (https://github.com/vmware-tanzu/carvel-ytt/pull/677#discussion_r910448304)
- [x] put new `@ytt:assert` functions into "validation" experiment to preserve the "right" to change the API.
  - [x] update docs for each such function with the same warning we have for `@schema/validation`
- [x] verify that duplicate validation kwargs are detected and result in an error.
- [x] switch the order of `fail()` and `True` for existing rules — to make the code more intuitive/readable.
- [x] `one_not_null()`/`one_not_null=`
  - [x] accept sequence of keys or bool for `assert.one_not_null()` and `one_not_null=`
  - [x] schema specific tests (`one_not_null=`)
  - [x] report helpful error message when value(s) is/are not a map.
- [x] Update GoDoc for `Max()`/`Min()`/`MaxLength()`/`MinLength()`
- [x] exploratory test edge-cases for inputs to validation kwargs.
  - when two validations rules contradict (e.g. min/max where max < min; min_len/max_len where max_len < min_len; etc.)
  - 0 cases
  - assert.<assertion-function>() with a misspelled name
  - where min=/max= are given a value that has no minimum or maximum defined, does the error message make sense? (e.g. None?)

## Fit and Finish

- [x] are there "TODO"s left in the code? Let's implement or remove.

### CODE REVIEW

File: pkg/validations/validate.go [hunk 1]
```diff
@@ -131,8 +132,10 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
 // shouldValidate uses validationKwargs and the node's value to run checks on the value. If the value satisfies the checks,
 // then the NodeValidation's rules should execute, otherwise the rules will be skipped.
 func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
-	_, isNone := value.(starlark.NoneType)
-	if isNone && (v.whenNullSkip != nil && *v.whenNullSkip) {
+	_, isNull := value.(starlark.NoneType)
+	nullAllowed := !v.notNull
+	whenNullSkip := v.whenNullSkip != nil && *v.whenNullSkip
+	if isNull && nullAllowed && whenNullSkip {
```
Comment (pivotaljohn) [inline pkg/validations/validate.go:138]: To keep this `if` from getting out of hand, captured each intention in a variable.

File: pkg/yttlibrary/assert.go [hunk 2]
```diff
@@ -261,21 +265,14 @@ func (b assertModule) Max(thread *starlark.Thread, f *starlark.Builtin, args sta
...
 func (b assertModule) NotNull(thread *starlark.Thread, f *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
-	if len(args) == 0 {
-		return starlark.None, fmt.Errorf("expected at least one argument.")
-	}
-	result, err := starlark.Call(thread, NewAssertNotNull(), args, []starlark.Tuple{})
-	if err != nil {
-		return starlark.None, err
-	}
-	return result, nil
+	return NewAssertNotNull(), nil
```
Comment (pivotaljohn) [inline pkg/yttlibrary/assert.go]: `assert.not_null()` is special: it does not take any parameters, so we just hand back the "assertion object", here. 👍🏻

File: pkg/validations/assert.go [hunk 3]
```diff
@@ -110,6 +119,31 @@ func newValidationKwargs(kwargs []starlark.Tuple, annPos *filepos.Position) (val
 			}
 			b := bool(v)
 			processedKwargs.whenNullSkip = &b
+		case ValidationKwargMinLength:
+			v, err := starlark.NumberToInt(value[1])
```
Comment (gcheadle-vmware) [inline pkg/validations/assert.go]: Should we document the rounding behavior of floats? or just document it as an int?

File: pkg/validations/assert.go [hunk 4]
```diff
@@ -110,6 +119,31 @@ func newValidationKwargs(kwargs []starlark.Tuple, annPos *filepos.Position) (val
 			}
 			b := bool(v)
 			processedKwargs.whenNullSkip = &b
+		case ValidationKwargMinLength:
```
Comment (gcheadle-vmware) [inline pkg/validations/assert.go]: If we consider someone providing multiple `min_len` kwargs, would it make sense to take the greater minimum value since they are comparable numbers?

However, it is consistent if we always took the last one, which is how all these kwargs are implemented.

File: pkg/validations/validate.go [hunk 5]
```diff
@@ -30,6 +32,12 @@ type rule struct {
 type validationKwargs struct {
 	when         *starlark.Callable
 	whenNullSkip *bool // default: nil if kwarg is not set, True if value is Nullable
+	minLength    int   // 0 len("") == 0, this always passes
+	maxLength    *int
+	min          starlark.Value
+	max          starlark.Value
+	notNull      bool
+	// *int , default=nil possible_values={&1, &2, &-3, ..}
```
Comment (gcheadle-vmware) [inline pkg/validations/validate.go]: should this comment be on line 36?

File: pkg/validations/validate.go [hunk 6]
```diff
@@ -125,8 +133,10 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
 // shouldValidate uses validationKwargs and the node's value to run checks on the value. If the value satisfies the checks,
 // then the NodeValidation's rules should execute, otherwise the rules will be skipped.
 func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
-	_, isNone := value.(starlark.NoneType)
-	if isNone && (v.whenNullSkip != nil && *v.whenNullSkip) {
+	_, isNull := value.(starlark.NoneType)
+	nullAllowed := !v.notNull
```
Comment (gcheadle-vmware) [inline pkg/validations/validate.go]: I have to remind myself that `v.notNull` is false by default in order to understand the if statement below.

File: pkg/validations/validate.go [hunk 7]
```diff
@@ -137,12 +147,53 @@ func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.
...
+		return isNull || isTrue, nil
 	}
 	// if no kwargs then execute rules
 	return true, nil
 }
 
+func (v validationKwargs) convertToRules() []rule {
+	var rules []rule
+	if minLen := v.minLength; minLen > 0 {
+		a := yttlibrary.NewAssertMinLen(core.NewGoValue(minLen).AsStarlarkValue())
```
Comment (gcheadle-vmware) [inline pkg/validations/validate.go]: I'm wondering if `minLen` and `maxLen` could just remain as their starlark types? It looks like we are converting it from a `starlark.Int` to an `int` when we construct `validationKwargs`, and then back to a starlark value here. We could save some sweat if we kept it as a starlark value the whole time.

The only concern I have with making such a change is that we are using the int in multiple places around this line of code (`minLen > 0`, `fmt.Sprintf("length greater or equal to %v", minLen)`).

**Review by gcheadle-vmware — COMMENTED** [review]
Overall, this PR is shaping up really nicely. Really proud of our group effort! @vmunishwar, @pivotaljohn 
I've left my additional thoughts in the comments.

File: pkg/validations/assert.go [hunk 8]
```diff
@@ -59,8 +64,8 @@ func (a *convertAssertAnnsToValidations) Visit(node yamlmeta.Node) error {
 // If any value in the annotation is not well-formed, it returns an error.
 func NewValidationFromValidationAnnotation(annotation template.NodeAnnotation) (*NodeValidation, error) {
 	var rules []rule
-
-	if len(annotation.Args) == 0 {
+	// TODO: what makes a malformed annotation??
```
Comment (gcheadle-vmware) [inline pkg/validations/assert.go]: Is this a malformed annotation? `#@schema/validation max_len=3, min_len=5`

File: pkg/validations/assert.go (hunk 3 above)
Comment (pivotaljohn) [reply pkg/validations/assert.go]: Documenting as an integer, but accepting any number seems legit.

Starlark is dynamically typed; programmers will expect automatic coercion of numbers (maybe not necessarily 1.5 => 1, but 1.0 => 1, for sure).

File: pkg/validations/assert.go (hunk 4 above)
Comment (pivotaljohn) [reply pkg/validations/assert.go]: I'm trying to come up with a use-case where an author _intentionally_ uses a keyword argument more than once...

... and the most likely scenario I can come up with is that they did it on accident ... which — if true — I bet they'd appreciate hearing about that.

So, you've got me starting to rethink the current behavior... that we'd rather error out on a duplicate keyword. 🤔 

Thoughts?

File: pkg/validations/validate.go (hunk 5 above)
Comment (pivotaljohn) [reply pkg/validations/validate.go]: yeah! wooops!

File: pkg/validations/validate.go (hunk 6 above)
Comment (pivotaljohn) [reply pkg/validations/validate.go]: excellent feedback. I see... `nullAllowed` has no apparent connection to the validation...  yeah, while there's a double negative (which is a little harder to reason) it's better than a full-on level of indirection. 👍🏻 

How about this?

```go
func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
	_, valueIsNull := value.(starlark.NoneType)
	whenNullSkip := v.whenNullSkip != nil && *v.whenNullSkip
	if !v.notNull && whenNullSkip && valueIsNull {
		return false, nil
	}
```

File: pkg/validations/validate.go (hunk 7 above)
Comment (pivotaljohn) [reply pkg/validations/validate.go]: It's a good question.

A big driver for `min` and `max` to become `starlark.Value`s is that many (but not all) types in Starlark can be compared: we needed the wider type, there.
`min_len` and `max_len` _only_ max sense as an integer.

But to your point, we _could_ store these values as `starlark.Int`s... which have a convenient accessor to the underlying int64 value (for the usages you pointed at).

Let's do that!

File: pkg/validations/validate.go (hunk 5 above)
Comment (pivotaljohn) [reply pkg/validations/validate.go]: Okay... conversation below about storing these values as `starlark.Int` obviates this comment all together. Let's delete.

File: pkg/validations/assert.go (hunk 8 above)
Comment (pivotaljohn) [reply pkg/validations/assert.go]: I'd say it's well-formed because it's syntactically correct.

However, it just won't work in practice. I have an assumption that library authors will test their business... and will discover that there's no set of data values that can satisfy all these rules.

File: pkg/validations/validate.go (hunk 5 above)
Comment (vmunishwar) [reply pkg/validations/validate.go]: Done. Good catch, Garrett!

**Comment by vmunishwar** [comment]
To be done later - review tests for duplicate coverage; every test represents maintenance, let's make sure they are doing their work.

File: pkg/yttlibrary/assert.go [hunk 9]
```diff
@@ -8,26 +8,51 @@ import (
...
+	members := starlark.StringDict{
+		"equals": starlark.NewBuiltin("assert.equals", core.ErrWrapper(m.Equals)),
+		"fail":   starlark.NewBuiltin("assert.fail", core.ErrWrapper(m.Fail)),
+		"try_to": starlark.NewBuiltin("assert.try_to", core.ErrWrapper(m.TryTo)),
+	}
+	if experiments.IsValidationsEnabled() {
+		members = starlark.StringDict{
+			"equals":       starlark.NewBuiltin("assert.equals", core.ErrWrapper(m.Equals)),
+			"fail":         starlark.NewBuiltin("assert.fail", core.ErrWrapper(m.Fail)),
+			"try_to":       starlark.NewBuiltin("assert.try_to", core.ErrWrapper(m.TryTo)),
```
Comment (cppforlife) [inline pkg/yttlibrary/assert.go]: instead of overriding members entirely, we should just be able to set new keys (keeping baseline the same)

File: pkg/validations/assert.go [hunk 10]
```diff
@@ -71,45 +76,109 @@ func NewValidationFromValidationAnnotation(annotation template.NodeAnnotation) (
...
 		message, ok := ruleTuple[0].(starlark.String)
 		if !ok {
 			return nil, fmt.Errorf("expected first item in the 2-tuple to be a string describing a valid value, but was %s (at %s)", ruleTuple[0].Type(), annotation.Position.AsCompactString())
 		}
-		lambda, ok := ruleTuple[1].(starlark.Callable)
+
+		assertion, ok := ruleTuple[1].(starlark.Callable)
 		if !ok {
-			return nil, fmt.Errorf("expected second item in the 2-tuple to be an assertion function, but was %s (at %s)", ruleTuple[1].Type(), annotation.Position.AsCompactString())
+			var err error
```
Comment (cppforlife) [inline pkg/validations/assert.go]: nit: unnecessary decl

File: pkg/validations/assert.go [hunk 11]
```diff
@@ -71,45 +76,109 @@ func NewValidationFromValidationAnnotation(annotation template.NodeAnnotation) (
...
+			case starlark.Bool:
+				if v {
+					processedKwargs.oneNotNull = v
+				} else {
+					return validationKwargs{}, fmt.Errorf("one_not_null= cannot be False")
+				}
+			case starlark.Sequence:
+				processedKwargs.oneNotNull = v
+			default:
+				return validationKwargs{}, fmt.Errorf("expected True or a sequence of keys, but was a \"%s\"", value[1].Type())
```
Comment (cppforlife) [inline pkg/validations/assert.go]: nit: we've typically used single quotes for user controlled values.

File: pkg/validations/validate.go [hunk 12]
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
 // shouldValidate uses validationKwargs and the node's value to run checks on the value. If the value satisfies the checks,
 // then the NodeValidation's rules should execute, otherwise the rules will be skipped.
 func (v validationKwargs) shouldValidate(value starlark.Value, thread *starlark.Thread) (bool, error) {
-	_, isNone := value.(starlark.NoneType)
-	if isNone && (v.whenNullSkip != nil && *v.whenNullSkip) {
+	_, valueIsNull := value.(starlark.NoneType)
+	whenNullSkip := v.whenNullSkip != nil && *v.whenNullSkip
+	if !v.notNull && whenNullSkip && valueIsNull {
```
Comment (cppforlife) [inline pkg/validations/validate.go]: todo do we really need `!v.notNull` since valueIsNull==true

File: pkg/validations/validate.go [hunk 13]
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
...
 		return false, nil
 	}
 
-	if v.when != nil {
-		result, err := starlark.Call(thread, *v.when, starlark.Tuple{value}, []starlark.Tuple{})
+	if !(v.when == nil || reflect.ValueOf(v.when).IsNil()) {
```
Comment (cppforlife) [inline pkg/validations/validate.go]: nit: more readable with ! + &&

File: pkg/validations/validate.go [hunk 14]
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
...
+		result, err := starlark.Call(thread, v.when, starlark.Tuple{value}, []starlark.Tuple{})
 		if err != nil {
 			return false, err
 		}
 
 		isTrue := bool(result.Truth())
-		return isNone || isTrue, nil
+		return valueIsNull || isTrue, nil
```
Comment (cppforlife) [inline pkg/validations/validate.go]: valueIsNull? are we saying we should always validate for null values.

File: pkg/validations/validate.go [hunk 15]
```diff
@@ -125,24 +133,90 @@ func (v *NodeValidation) DefaultNullSkipTrue() {
...
 	}
 	// if no kwargs then execute rules
 	return true, nil
 }
 
+func (v validationKwargs) convertToRules() []rule {
+	var rules []rule
+
+	if v.minLength != nil {
+		a := yttlibrary.NewAssertMinLen(*v.minLength)
```
Comment (cppforlife) [inline pkg/validations/validate.go]: nit: inline assertion constuctor

File: pkg/yttlibrary/assert.go [hunk 16]
```diff
@@ -101,3 +129,289 @@ func (b assertModule) TryTo(thread *starlark.Thread, f *starlark.Builtin, args s
 	}
 	return starlark.Tuple{retVal, starlark.None}, nil
 }
+
+// Assertion encapsulates a rule (a predicate) that can be accessed in a Starlark expression (via the "check" attribute)
+// or in Go (via CheckFunc()).
+type Assertion struct {
```
Comment (cppforlife) [inline pkg/yttlibrary/assert.go:129]: lets add ConversionHint

File: pkg/validations/assert.go (hunk 10 above)
Comment (pivotaljohn) [reply pkg/validations/assert.go]: In fact, the companion variable, `assertion` has a wider scope so this `var` is at least necessary, if not undesirable.

The scope of `err`, here, is only within this `if`, so seems appropriate.

File: pkg/validations/assert.go (hunk 11 above)
Comment (pivotaljohn) [reply pkg/validations/assert.go]: There were a couple other places where this slipped in as well; fixed those too.

File: pkg/validations/validate.go (hunk 12 above)
Comment (pivotaljohn) [reply pkg/validations/validate.go]: Turns out: we do.

`!v.notNull` should be read, "is null allowed?" ... so we should only entertain skipping validating this value if null is allowed.

introduced a variable to improve the readability, here.

File: pkg/validations/validate.go (hunk 15 above)
Comment (pivotaljohn) [reply pkg/validations/validate.go]: ... and remaining `a` expanded to the more selectable, `assertion`. 👍🏻
